internal/resources/kubernetessecret/data_source_secret.go

```diff
@@ -0,0 +1,108 @@
+	clienterrors "github.com/vmware/terraform-provider-tanzu-mission-control/internal/client/errors"
+	"github.com/vmware/terraform-provider-tanzu-mission-control/internal/helper"
+	secretclustermodel "github.com/vmware/terraform-provider-tanzu-mission-control/internal/models/kubernetessecret/cluster"
+	secretexportclustermodel "github.com/vmware/terraform-provider-tanzu-mission-control/internal/models/kubernetessecret/cluster/secretexport"
+	"github.com/vmware/terraform-provider-tanzu-mission-control/internal/resources/common"
+)
+
+func DataSourceSecret() *schema.Resource {
+	return &schema.Resource{
+		ReadContext: func(ctx context.Context, d *schema.ResourceData, m interface{}) diag.Diagnostics {
+			return dataSourceSecretRead(context.WithValue(ctx, contextMethodKey{}, "dataSourceRead"), d, m)
+		},
+		Schema: secretSchema,
+	}
+}
+
+func dataSourceSecretRead(ctx context.Context, d *schema.ResourceData, m interface{}) diag.Diagnostics {
+	config := m.(authctx.TanzuContext)
+
+	var (
+		diags            diag.Diagnostics
+		secretResp       *secretclustermodel.VmwareTanzuManageV1alpha1ClusterNamespaceGetSecretResponse
+		secretexportResp *secretexportclustermodel.VmwareTanzuManageV1alpha1ClusterNamespaceGetSecretExportResponse
+		err              error
+	)
+
+	secretName, _ := d.Get(NameKey).(string)
+
+	secretResp, err = config.TMCConnection.SecretResourceService.SecretResourceServiceGet(constructFullname(d))
+	if err != nil || secretResp == nil {
+		if clienterrors.IsNotFoundError(err) {
+			if ctx.Value(contextMethodKey{}) == "dataSourceRead" {
+				return diag.FromErr(errors.Wrapf(err, "Tanzu Mission Control cluster secret entry not found, name : %s", d.Get(NameKey)))
+			}
+
+			_ = schema.RemoveFromState(d, m)
+
+			return diags
+		}
+
+		return diag.FromErr(errors.Wrapf(err, "unable to get Tanzu Mission Control secret entry, name : %s", secretName))
+	}
+
+	var pass string
+
+	if _, ok := d.GetOk(specKey); ok {
+		pass, _ = (d.Get(helper.GetFirstElementOf(specKey, DockerConfigjsonKey, PasswordKey))).(string)
+	}
+
+	status := map[string]interface{}{
+		SecretPhaseKey: secretResp.Secret.Status.Conditions["Ready"].Reason,
+	}
+
+	if d.Get(ExportKey).(bool) {
+		secretexportResp, err = config.TMCConnection.SecretExportResourceService.SecretExportResourceServiceGet(constructFullnameSecetExport(d))
+		if err != nil || secretexportResp == nil {
+			switch {
+			case clienterrors.IsNotFoundError(err):
+				if err := d.Set(ExportKey, false); err != nil {
+					return diag.FromErr(err)
+				}
+			default:
+				return diag.FromErr(errors.Wrapf(err, "unable to get Tanzu Mission Control secret export entry, name : %s", secretName))
+			}
+		} else {
+			status[SecretExportPhaseKey] = secretexportResp.SecretExport.Status.Conditions["Ready"].Reason
+		}
+	} else {
+		secretexportResp, err = config.TMCConnection.SecretExportResourceService.SecretExportResourceServiceGet(constructFullnameSecetExport(d))
```

Comment:
why is the call being done in if and else ?

Reply:
update PTAL